src/main/java/org/icatproject/core/manager/search/OpensearchApi.java

```diff
@@ -0,0 +1,1725 @@
+		// and know their parent's id (N.B. InvestigationUsers are also mapped to
+		// Datasets and Datafiles, but using the investigation.id field)
+		relations.put("datafileparameter", Arrays.asList(
+				new ParentRelation(RelationType.NESTED_CHILD, "datafile", "datafile", null)));
+		relations.put("datasetparameter", Arrays.asList(
+				new ParentRelation(RelationType.NESTED_CHILD, "dataset", "dataset", null)));
+		relations.put("datasettechnique", Arrays.asList(
+				new ParentRelation(RelationType.NESTED_CHILD, "dataset", "dataset", null)));
+		relations.put("investigationparameter", Arrays.asList(
+				new ParentRelation(RelationType.NESTED_CHILD, "investigation", "investigation", null)));
+		relations.put("sampleparameter", Arrays.asList(
+				new ParentRelation(RelationType.NESTED_CHILD, "investigation", "sample", null), // Must be first
+				new ParentRelation(RelationType.NESTED_CHILD, "dataset", "sample", null),
+				new ParentRelation(RelationType.NESTED_CHILD, "datafile", "sample", null)));
+		relations.put("investigationuser", Arrays.asList(
+				new ParentRelation(RelationType.NESTED_CHILD, "investigation", "investigation", null),
+				new ParentRelation(RelationType.NESTED_CHILD, "dataset", "investigation", null),
+				new ParentRelation(RelationType.NESTED_CHILD, "datafile", "investigation", null)));
+		relations.put("investigationinstrument", Arrays.asList(
+				new ParentRelation(RelationType.NESTED_CHILD, "investigation", "investigation", null),
+				new ParentRelation(RelationType.NESTED_CHILD, "dataset", "investigation", null),
+				new ParentRelation(RelationType.NESTED_CHILD, "datafile", "investigation", null)));
+
+		// Grandchildren are entities that are related to one of the nested
+		// children, but do not have a direct reference to one of the indexed entities,
+		// and so must be updated by query - they also only affect a subset of the
+		// nested fields, rather than an entire nested object
+		relations.put("parametertype", Arrays.asList(
+				new ParentRelation(RelationType.NESTED_GRANDCHILD, "investigation", "investigationparameter",
+						ParameterType.docFields),
+				new ParentRelation(RelationType.NESTED_GRANDCHILD, "investigation", "sampleparameter",
+						ParameterType.docFields),
+				new ParentRelation(RelationType.NESTED_GRANDCHILD, "dataset", "datasetparameter",
+						ParameterType.docFields),
+				new ParentRelation(RelationType.NESTED_GRANDCHILD, "dataset", "sampleparameter",
+						ParameterType.docFields),
+				new ParentRelation(RelationType.NESTED_GRANDCHILD, "datafile", "datafileparameter",
+						ParameterType.docFields),
+				new ParentRelation(RelationType.NESTED_GRANDCHILD, "datafile", "sampleparameter",
+						ParameterType.docFields)));
+		relations.put("technique", Arrays.asList(
+				new ParentRelation(RelationType.NESTED_GRANDCHILD, "dataset", "datasettechnique",
+						Technique.docFields)));
+		relations.put("user", Arrays.asList(
```

Comment:
That's a really good spot, here we should only have `"user"` as a key once and just have all 4 relations in one list (rather than a list of 1 and a list of 3).